lib/charms/vault_k8s/v0/vault_managers.py

```diff
@@ -504,3 +526,288 @@ def generate_vault_unit_certificate(
+    def mount_path(self) -> str:
+        """Return the mount path for the transit backend."""
+        return self._mount_path
+
+    def sync(self) -> None:
+        """Ensure that all auto-unseal requests are fulfilled and clean up unused credentials.
+
+        This looks for any outstanding requests for auto-unseal that may have
+        been missed. If there are any, it generates the credentials and sets
+        them in the relation databag.
+
+        It also cleans up any credentials that are no longer used by any of the
+        relations, and logs a warning about orphaned keys. It will not remove
+        any keys, to prevent loss of data.
+        """
+        if not self._juju_facade.is_leader:
+            return
+        outstanding_requests = self._provides.get_outstanding_requests()
+        if outstanding_requests:
+            self._client.ensure_secrets_engine(SecretsBackend.TRANSIT, self._mount_path)
+        for relation in outstanding_requests:
+            self.create_credentials(relation)
+
+        self._clean_up_credentials()
+
+    def _clean_up_credentials(self) -> None:
+        """Clean up roles and policies that are no longer needed by autounseal.
+
+        This method will remove any roles and policies that are no longer
+        used by any of the existing relations. It will also detect any orphaned
+        keys (keys that are not associated with any relation) and log a warning.
+        """
+        self._clean_up_roles()
+        self._clean_up_policies()
+        self._detect_and_allow_deletion_of_orphaned_keys()
+
+    def _detect_and_allow_deletion_of_orphaned_keys(self) -> None:
+        existing_keys = self._get_existing_keys()
+        relation_key_names = [
+            VaultAutounsealRelationDetails(relation).key_name
+            for relation in self._juju_facade.get_active_relations(self._provides.relation_name)
+        ]
+        orphaned_keys = [key for key in existing_keys if key not in relation_key_names]
+        if not orphaned_keys:
+            return
+        logging.warning(
+            f"Orphaned autounseal keys were detected: {orphaned_keys}. If you are sure these are no longer needed, you may manually delete them using the vault CLI to suppress this message."
+            " To delete a key, use the command `vault delete charm-autounseal/keys/<key_name>`."
+        )
+        for key_name in orphaned_keys:
+            deletion_allowed = self._is_deletion_allowed(key_name)
+            if not deletion_allowed:
+                self._allow_key_deletion(key_name)
+
+    def _allow_key_deletion(self, key_name) -> None:
+        logger.info("Allowing deletion of key %s", key_name)
+        self._client.write(f"{self.mount_path}/keys/{key_name}/config", {"deletion_allowed": True})
+
+    def _is_deletion_allowed(self, key_name) -> bool:
+        data = self._client.read(f"{self.mount_path}/keys/{key_name}")
+        return data["deletion_allowed"]
+
+    def _clean_up_roles(self) -> None:
```

Comment:
I feel docstrings would be useful here to understand what roles we are deleting exactly and why. The same comment likely applies to policies. Example `Delete autounseal Vault roles that are not associated with an existing autounseal Juju integration`

Reply:
https://github.com/canonical/vault-k8s-operator/pull/534/commits/8a2350e3313db1a915d845bdac3ef778d86e4ec4